// rmi/src/client/SbisEnv/_Maintains/Parking/Const.ts
/// <amd-module name="SbisEnv/_Maintains/Parking/Const" />
export const enum Size {
    large = 'large',
    medium = 'medium',
    normal = '',
}
export const MESSAGE = {
    REMOVED_TASK: "Сожалеем, но эта задача была удалена",
    REMOVED_DIALOG: "Сожалеем, но этот диалог был удален",
    REMOVED_DOCUMENT: "Сожалеем, но этот документ был удален",
    NOT_FOUND: "Сожалеем, вы перешли по  неверной ссылке",
    ACCESS_DENIED: "У вас нет доступа к этой странице",
    ACCESS_DENIED_TASK: "У вас нет доступа к этой странице",
    ACCESS_DENIED_DOCUMENT: "У вас нет прав на доступ к документу",
    ACCESS_DENIED_MESSAGE: "К сожалению, диалог вам недоступе",
    SERVICE_FOR_MAINTENANCE: "Извините, мы на техническом обслуживании",
    ACCOUNT_MANAGEMENT: "Идет модернизация кабинета",
    CONNECTION_ERROR: "Произошел разрыв соединения",
    UPDATE: "СБИС обновляется до версии",
    INTERNAL: "У СБИС проблема"
};
export const ACTION = {
    RELOAD: "Обновить страницу",
    MAIN_PAGE: "Перейти на главную",
    ASK_ACCESS_AUTHOR: "Отправить запрос его автору",
    REQUESTED: "Запрос отправлен, ожидайте",
    ASK_ACCESS_GROUP: "Отправьте запрос его участникам",
    ASK_SUPPORT: "Обратитесь в техподдержку",
    ASK_ADMIN: "администратору системы"
};
export const DETAILS = {
    ASK_ADMIN: "Обратитесь к ",
    ACCOUNT_MANAGEMENT: "Скоро вам будут доступны дополнительные возможности СБИС",
    CONNECTION_ERROR: "Проверьте подключение к интернету",
    INTERNAL: "",
    NOT_FOUND: "",
    SERVICE_FOR_MAINTENANCE: ""
};
export const IMAGE = {
    NOT_FOUND: 'https://i.pinimg.com/474x/54/5a/0f/545a0f60746c7a8eeeb396082c768952.jpg',
    SERVICE_FOR_MAINTENANCE: 'https://i.pinimg.com/474x/54/5a/0f/545a0f60746c7a8eeeb396082c768952.jpg',
    INTERNAL: 'https://i.pinimg.com/474x/54/5a/0f/545a0f60746c7a8eeeb396082c768952.jpg',
    CONNECTION_ERROR: 'https://i.pinimg.com/474x/54/5a/0f/545a0f60746c7a8eeeb396082c768952.jpg',
    ACCESS_DENIED: 'https://i.pinimg.com/474x/54/5a/0f/545a0f60746c7a8eeeb396082c768952.jpg',
    // TODO: отдельное лого для каждой локализации
    HEADER_LOGO: 'http://axure.tensor.ru/Platform_8/images//заглушки_3_0/regen/u15.png',
};
