// rmi/src/client/SbisEnv/_Maintains/Parking/_handlers/accessDenied.ts
/// <amd-module name="SbisEnv/_Maintains/Parking/_handlers/accessDenied" />
import { MESSAGE, IMAGE, DETAILS } from "SbisEnv/_Maintains/Parking/Const";
import Config from "SbisEnv/_Maintains/Parking/Config";
import { Handler, Mode } from "SbisEnv/Error";
// @ts-ignore
import { HTTP } from 'Transport/Errors';
import Body from 'SbisEnv/_Maintains/Parking/Body';
import getDefaultOptions from "SbisEnv/_Maintains/Parking/_handlers/getDefOptions";
import { MainPage, AskAdmin } from 'SbisEnv/_Maintains/Parking/action';
import Page from 'SbisEnv/_Maintains/Parking/Page'
let accessDenied: Handler<HTTP, Config> = ({ error, mode }) => {
    if (error.code != 403) {
        return ;
    }
    return {
        template: mode == Mode.page ? Page: Body,
        options: {
            ...getDefaultOptions(error, mode),
            message: rk(MESSAGE.ACCESS_DENIED),
            image: IMAGE.ACCESS_DENIED,
            details: AskAdmin,
            action: MainPage
        }
    }
};

export default accessDenied;
